### ISSUE TITLE
Using MeshGraphNets on own data?

### ISSUE THREAD

**Issue by mtbc** [deepmind/deepmind-research#267]
We have training sets with time steps of triangular meshes of the surface of volumes that deform over time and are hoping that your MeshGraphNets approach for Lagrangian systems may be a big win over our current finite element approach which takes days to generate predictions. The meshing is dynamic but it seems that you do not require the same nodes to exist in each timestep.

I have had some success in trying your code. I know some Python but not TensorFlow. I managed to get the older Python 3.6 and numpy 1.18.5 installed so I could run your code with TensorFlow 1, though your `run.sh` seems to do things then delete it all without showing anything, even if I also install `tk` into the virtual environment. TensorFlow record datasets are new to me but, even if I learn about that, I don't know if it's clear what the input and output formats of your software are within those datasets. At least the code itself seems somewhat clear, much appreciated.

I guess, I'm faced with a dilemma with which I'd appreciate your thoughts. We have limited time and funds and could:

* If you think that your code would be a good fit for our surface deformation prediction problem, we could try to understand it all better and apply it directly and let you know how that goes. We'd certainly appreciate any clues along the way in terms of input or output formats, what kind of model to apply (cloth?), if we must do something special to account for the dynamic meshing, etc. I don't know if any of your demo datasets are already very similar, you seem to provide more than are described in the paper.

* I think that your approach may be great, especially as we can get meshes but, depending on the difficulty of our applying your software, it's also possible that it might be easiest for us to not try to use it directly, instead coding something based on your ideas and reusing specific parameters (network size and depth, etc.) wherever you have left us some low-hanging hard-won fruit. We'd possibly use PyTorch instead as we are already using it in trying a voxel/image-based approach in applying U-Nets to the same problem.

Any advice or pointers are appreciated. I should note that this issue may be related to #199 and #204.

**Comment by alvarosg** [comment]
Thank you for your questions! 

> would be a good fit for our surface deformation prediction problem
> we could try to understand it all better and apply it directly and let you know how that goes
> We'd possibly use PyTorch instead as we are already using it

For the time being, we provided the reference implementation to accompany the paper for illustration purposes, but there are no plans to maintain the implementation or add new features. If you are not familiar with TF1, my recommendation would be to use the implementation as reference, and then work with the framework that you are most familiar with. You could always run one of the existing datasets in both our previous and your existing implementation.

Of course, you could give it a try to create a dataset with the same format and try to run it. Not it is not mandatory to use "tf.records" to store the dataset. `tf.data.Dataset` has a [from_generator](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#from_generator), that lets you load a dataset using arbitrary python code, so you may read your dataset in whatever format you have it stored and simply make the tensors look like those in our tf records datasets.

Note the current example only covers the simpler case of the cloth, which does not add additional collision edges (it just uses the mesh edges), so if your problem requires that (e.g. in case of contact forces), you would have to create a model with an additional collisions edges edge set, rather than a single mesh edges set like in [the demo](https://github.com/deepmind/deepmind-research/blob/30799687edb1abca4953aec507be87ebe63e432d/meshgraphnets/cloth_model.py#L65).

> though your run.sh seems to do things then delete it all without showing anything

run.sh is continuous integration test, which is why it does clean up after it finishes. If you want to train a model and visualize outputs please follow the instructions in the README :)

Hope this helps!

**Comment by mtbc** [comment]
Thank you, much appreciated, especially that I can at least ask questions about the existing implementation. Happy to keep this issue as a record of what we try when but, as it'll be a slow burn on our side, might be better from your side if we just close this issue and I open any new one for any specific things about your code as I come to them?

At least we indeed don't have any collisions to think about, just irregularly shaped regions somewhat changing shape, so it could be that the current example you provide is already close to fitting our work.

**Comment by yq60523** [comment]
I have some questions about the data structure of the meshgraphnet, especially for CylinderFlow and Airfoil:
cells: [num_steps, num_nodes, num_dimensions(3, triangular mesh is combined with three nodes)]
mesh-pos: [num_steps, num_nodes, num_dimensions(2)]
node_type: [num_steps, num_nodes, num_dimensions(1)]
velocity: [num_steps, num_nodes, num_dimensions(2, x-direction velocity and y-direction velocity of each node)]
pressure: [num_steps, num_nodes, num_dimensions(1, the pressure of each node)]

I can't make sure my understanding of your dataset is true. So, if my understanding is wrong, please point it out. Thanks.

There are still some questions else existing.
1. For the node_type: can you explain the node type carefully? such as the meaning of NORMAL, OBSTACLE, AIRFOIL, HANDLE, INFLOW, OUTFLOW, WALL_BOUNDARY and SIZE?
2. The data of the velocity and pressure is the node's velocity and pressure, not the grid's, is it?

Looking forward to your reply. Thanks.

**Comment by yq60523** [comment]
> Thank you for your questions!
> 
> > would be a good fit for our surface deformation prediction problem
> > we could try to understand it all better and apply it directly and let you know how that goes
> > We'd possibly use PyTorch instead as we are already using it
> 
> For the time being, we provided the reference implementation to accompany the paper for illustration purposes, but there are no plans to maintain the implementation or add new features. If you are not familiar with TF1, my recommendation would be to use the implementation as reference, and then work with the framework that you are most familiar with. You could always run one of the existing datasets in both our previous and your existing implementation.
> 
> Of course, you could give it a try to create a dataset with the same format and try to run it. Not it is not mandatory to use "tf.records" to store the dataset. `tf.data.Dataset` has a [from_generator](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#from_generator), that lets you load a dataset using arbitrary python code, so you may read your dataset in whatever format you have it stored and simply make the tensors look like those in our tf records datasets.
> 
> Note the current example only covers the simpler case of the cloth, which does not add additional collision edges (it just uses the mesh edges), so if your problem requires that (e.g. in case of contact forces), you would have to create a model with an additional collisions edges edge set, rather than a single mesh edges set like in [the demo](https://github.com/deepmind/deepmind-research/blob/30799687edb1abca4953aec507be87ebe63e432d/meshgraphnets/cloth_model.py#L65).
> 
> > though your run.sh seems to do things then delete it all without showing anything
> 
> run.sh is continuous integration test, which is why it does clean up after it finishes. If you want to train a model and visualize outputs please follow the instructions in the README :)
> 
> Hope this helps!

Hi, Alvaro.
I’ve generated my own data and also run it successfully. But I encountered a problem, when I want to add a new dimension as node features, such as solid concentration. The network_output will return Nan and the loss is also Nan. Do any other hyper parameters need to be modified?

**Comment by alvarosg** [comment]
If this is just an input feature that does not need to be predicted and you are adding it similarly to the other input features such as node type, then it should work. If you are getting nans, it is probably because of a different reason (perhaps the feature has a very bad scale). Is the loss nan even for the very first iteration?

If it is not the scale, or something like that, then I am afraid it will be very hard for us to say what me be wrong.

**Comment by yq60523** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by yq60523** [comment]
> If this is just an input feature that does not need to be predicted and you are adding it similarly to the other input features such as node type, then it should work. If you are getting nans, it is probably because of a different reason (perhaps the feature has a very bad scale). Is the loss nan even for the very first iteration?
> 
> If it is not the scale, or something like that, then I am afraid it will be very hard for us to say what me be wrong.

Yeah, it happens in the first iteration.
When I debug it, this error is caused by the minus variance in the normalization of the solid concentration, leading to a Nan standard deviation. I guess the cause of this error is the python float number mechanism. For example, when I input 0.0414, the actual number obtained by the python is 0.04140000005.
